crates/bootstrap_srv/src/server.rs

```diff
@@ -0,0 +1,333 @@
+            let store = store.clone();
+            let server = server.clone();
+            let space_map = space_map.clone();
+            workers.push(std::thread::spawn(move || {
+                worker(config, cont, store, server, space_map)
+            }));
+        }
+
+        // also set up a worker for pruning expired infos
+        let maint_cont = cont.clone();
+        let maint_space_map = space_map.clone();
+        workers.push(std::thread::spawn(move || {
+            maint_worker(config, maint_cont, maint_space_map)
+        }));
+
+        Ok(Self {
+            cont,
+            workers,
+            addr,
+        })
+    }
+
+    /// Get the bound listinging address of this server.
+    pub fn listen_addr(&self) -> std::net::SocketAddr {
+        self.addr
+    }
+}
+
+fn maint_worker(
+    config: Arc<Config>,
+    cont: Arc<std::sync::atomic::AtomicBool>,
+    space_map: crate::SpaceMap,
+) -> std::io::Result<()> {
+    let mut last_check = std::time::Instant::now();
+
+    while cont.load(std::sync::atomic::Ordering::SeqCst) {
+        std::thread::sleep(config.request_listen_duration);
+
+        if last_check.elapsed() >= config.prune_interval {
+            last_check = std::time::Instant::now();
+
+            space_map.update_all(config.max_entries_per_space);
+        }
+    }
+
+    Ok(())
+}
+
+fn worker(
+    config: Arc<Config>,
+    cont: Arc<std::sync::atomic::AtomicBool>,
+    store: Arc<crate::Store>,
+    server: Arc<Server>,
+    space_map: crate::SpaceMap,
+) -> std::io::Result<()> {
+    while cont.load(std::sync::atomic::Ordering::SeqCst) {
+        let req = match server.recv_timeout(config.request_listen_duration)? {
+            Some(req) => req,
+            None => continue,
+        };
+
+        let path = req
+            .url()
+            .split('/')
+            .rev()
+            .filter_map(|p| {
+                if p.is_empty() {
+                    None
+                } else {
+                    Some(p.to_string())
+                }
+            })
+            .collect::<Vec<_>>();
+
+        let handler = Handler {
+            config: &config,
+            store: &store,
+            space_map: &space_map,
+            method: req.method().as_str().to_string(),
+            path,
+            req,
+        };
+
+        handler.handle()?;
+    }
+    Ok(())
+}
+
+struct Handler<'lt> {
+    config: &'lt Config,
+    store: &'lt crate::Store,
+    space_map: &'lt crate::SpaceMap,
+    method: String,
+    path: Vec<String>,
+    req: tiny_http::Request,
+}
+
+impl<'lt> Handler<'lt> {
+    /// Wrap the handle call so we can respond to the client with errors.
+    pub fn handle(mut self) -> std::io::Result<()> {
+        match self.handle_inner() {
+            Ok((status, body)) => self.respond(status, body),
+            Err(err) => self.respond(
+                500,
+                serde_json::to_string(&serde_json::json!({
+                    "error": format!("{err:?}"),
```

Comment:
It looks like there's only one error where the error string is more than just a "type" label (line 259 `                std::io::Error::other(format!("InvalidSignature: {err:?}"))`.

If we want to include any message along with a "type" label, then I'd want them split them out into 2 fields for clients to easily match on.

Reply:
@mattyg - For this and https://github.com/holochain/kitsune2/pull/10#discussion_r1854225991 (which I don't consider a nitpick, I agree these are errors in user input, not in what the server is doing), I wonder if the error reporting types, since they are defined in json, and statuses actually belong in the spec itself. Would you be okay with updating the spec and fixing these responses being a follow-on effort after these are merged?

Reply:
[Created Issue](https://github.com/holochain/kitsune2/issues/17)